pkg/lifecycle/resource_debug_enabled.go

```diff
@@ -0,0 +1,115 @@
+// +build debug_ref
+
+package lifecycle
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"syscall"
+)
+
+// When in debug mode, we associate each reference an id and with that id the
+// stack trace that created it. We can't directly refer to the reference here
+// because we also associate a finalizer to print to stderr if a reference is
+// leaked, including where it came from if possible.
+
+func resourceClosed() error {
+	var buf [4096]byte
+	return fmt.Errorf("resource closed:\n%s", buf[:runtime.Stack(buf[:], false)])
+}
+
+// liveReferences keeps track of the stack traces of all of the live references.
+type liveReferences struct {
+	mu   sync.Mutex
+	id   uint64
+	live map[uint64][]uintptr
+}
+
+var live = &liveReferences{
+	live: make(map[uint64][]uintptr),
+}
+
+// finishId informs the liveReferences that the id is no longer in use.
+func (l *liveReferences) untrack(r *Reference) {
+	l.mu.Lock()
+	if _, ok := l.live[r.id]; ok {
+		delete(l.live, r.id)
+		runtime.SetFinalizer(r, nil)
+	}
```

Comment:
Is it worth checking `!ok`?

Reply:
I think I might get rid of the checking in general since the parent ensures that it's only called once and on tracked references, and the protected calls are idempotent/no-ops on untracked references.